src/proxy.rs

```diff
@@ -502,11 +527,129 @@ fn handle_server_event(
             );
 
             // TODO(#145): Fix unwrap
-            let handle = client_to_proxy.idle_reject(status).unwrap();
+            let handle = client_to_proxy.state.idle_reject(status).unwrap();
             trace!(role = "p2c", ?handle, "idle_reject");
         }
         client::Event::IdleDoneSent { handle } => {
             trace!(role = "p2s", ?handle, "--->");
         }
     }
 }
+
+struct ImapState<S> {
+    input_role: &'static str,
+    input_direction: &'static str,
+    input_color: Option<Color>,
+    input_fragmentizer: Fragmentizer,
+    output_role: &'static str,
+    output_direction: &'static str,
+    output_color: Option<Color>,
+    output_fragmentizer: Fragmentizer,
+    state: S,
+}
+
+impl ImapState<Client> {
+    pub fn client(state: Client) -> Self {
+        Self {
+            input_role: "s2p",
+            input_direction: "<--|",
+            input_color: Some(Color::Blue),
+            input_fragmentizer: Fragmentizer::without_max_message_size(),
+            output_role: "p2s",
+            output_direction: "--->",
+            output_color: None,
+            output_fragmentizer: Fragmentizer::without_max_message_size(),
+            state,
+        }
+    }
+}
+
+impl ImapState<Server> {
+    pub fn server(state: Server) -> Self {
+        Self {
+            input_role: "c2p",
+            input_direction: "|-->",
+            input_color: Some(Color::Red),
+            input_fragmentizer: Fragmentizer::without_max_message_size(),
+            output_role: "p2c",
+            output_direction: "<---",
+            output_color: None,
+            output_fragmentizer: Fragmentizer::without_max_message_size(),
+            state,
+        }
+    }
+}
+
+impl<S: imap_next::State> imap_next::State for ImapState<S> {
+    type Event = S::Event;
+    type Error = S::Error;
+
+    fn enqueue_input(&mut self, bytes: &[u8]) {
```

Comment:
question: Did you use something different than `enqueue_bytes` on purpose? We could call it `enqueue_bytes` as in `Fragmentizer`.

Reply:
`imap_next::Io` has two variants, `Io::NeedMoreInput` and `Io::Output`. The current name `State::enqueue_input` makes it clear that it needs to be called after `Io::NeedMoreInput` had occurred. However, no strong feelings if you want to change it.

Reply:
Let's keep it then :+1: